test/Microsoft.AspNet.Mvc.Test/MvcServiceCollectionExtensionsTest.cs

```diff
@@ -79,6 +83,71 @@ public void WithControllersAsServices_ScansControllersFromSpecifiedAssemblies()
             Assert.Equal(ServiceLifetime.Singleton, services[3].Lifetime);
         }
 
+        [Fact]
+        public void MultiRegistrationServiceTypes_AreRegistered_MultipleTimes()
+        {
+            // Arrange
+            var services = new ServiceCollection();
+            var multiRegistrationServiceTypes = GetMutliRegistrationServiceTypes();
+
+            // Act
+            MvcServiceCollectionExtensions.AddMvcServices(services);
+            MvcServiceCollectionExtensions.AddMvcServices(services);
+
+            // Assert
+            foreach (var serviceType in multiRegistrationServiceTypes)
+            {
+                AssertService(services, serviceType, 2);
+            }
+        }
+
+        [Fact]
+        public void SingleRegistrationServiceType_IsNotRegistered_MultipleTimes()
+        {
+            // Arrange
+            var services = new ServiceCollection();
+            var multiRegistrationServiceTypes = GetMutliRegistrationServiceTypes();
+
+            // Act
+            MvcServiceCollectionExtensions.AddMvcServices(services);
+            MvcServiceCollectionExtensions.AddMvcServices(services);
+
+            // Assert
+            var singleRegistrationServiceTypes = services
+                .Where(sd => !multiRegistrationServiceTypes.Contains(sd.ServiceType))
+                .Select(sd => sd.ServiceType);
+
+            foreach (var singleRegistrationType in singleRegistrationServiceTypes)
+            {
+                AssertService(services, singleRegistrationType, 1);
+            }
+        }
+
+        private IEnumerable<Type> GetMutliRegistrationServiceTypes()
+        {
+            return new[]
+            {
+                typeof(IConfigureOptions<MvcOptions>),
+                typeof(IConfigureOptions<RazorViewEngineOptions>),
+                typeof(IActionConstraintProvider),
+                typeof(IActionDescriptorProvider),
+                typeof(IActionInvokerProvider),
+                typeof(IFilterProvider),
+                typeof(IApiDescriptionProvider)
+            };
+        }
+
+        private void AssertService(IServiceCollection services, Type serviceType, int expectedServiceRegistrationCount)
+        {
+            var desc = services.Where(sd => sd.ServiceType == serviceType);
+            var actual = desc.Count();
+
+            Assert.True(
```

Comment:
I initially tried to do that but I wanted a clear message on which servicetype the asser failed. `Assert.Equals` doesn't have an overload taking in custom message.